[dnf5] libdnf/PackageSack: API for setting excludes / includes

m-blaha opened the pull request.

m-blaha committed:
libdnf/PackageSack: API for setting user excluded packages

j-mracek commented on libdnf/rpm/package_sack.cpp:
```diff
@@ -187,6 +188,14 @@ void PackageSack::Impl::setup_excludes_includes(bool only_main) {
     }
 }
 
+void PackageSack::Impl::add_excludes(const PackageSet & excludes) {
```
When this method is added then we have to ensure that `add_excludes()` and `setup_excludes_includes()` can be used in any order.

m-blaha replied:
Good catch - currently `setup_excludes_includes` resets excludes only in case that there are some defined in configuration. In case there is no exclude in configuration, the method doesn't do anything.

m-blaha replied:
I've changed `setup_excludes_includes` to always reset excludes.

j-mracek replied:
Be honest, I am not sure whether reset is a good idea. In case of reset we have to force API user to call methods in certain order.

m-blaha replied:
Reset in the sense of set them according the config files - which it didn't in case the config files did not contain any `excludepkgs=...` field.
Well, of course they need to call it in certain order - first setup_excludes_includes to (re)set excludes according the configuration and then use custom additions / removals.
It's true that name `setup_excludes_includes` is kind of misleading. Something like `read_excludes_includes_from_config` would be more descriptive, but also awfully long :)

m-blaha replied:
The other solution may be to adjust `setup_excludes_includes` to add excludes from configuration instead of setting it. Than the call order would not matter.

m-blaha replied:
I've added a commit that renames setup_excludes_includes to add_excludes_includes_from_config and changes behaviour accordingly. Now the methods can be called in any order.

j-mracek replied:
Sounds good to me, thank you.

j-mracek commented on libdnf/rpm/package_sack.cpp:
```diff
@@ -187,6 +188,14 @@ void PackageSack::Impl::setup_excludes_includes(bool only_main) {
 }
 
+void PackageSack::Impl::add_excludes(const PackageSet & excludes) {
+    if (&(pkg_excludes.get()->get_map()) == nullptr) {
```
`if (!pkg_excludes)` Is may be what you are looking for. What about to modify it to:
```
if (!pkg_excludes) {
    pkg_excludes.reset(new libdnf::solv::SolvMap(*excludes.p_impl));
} else {
    *pkg_excludes |= *excludes.p_impl;
}
considered_uptodate = false;
```

m-blaha committed:
libdnf/PackageSack: API for setting included packages

m-blaha committed:
Unit tests for excludes/includes API

lukash commented:
I think the user-set excludes should be stored separately and have an API along the lines of `get_user_excludes()`, `add_user_excludes()` and `clear_user_excludes()` (possibly also `set_user_excludes()`). Then, ideally, we'd lazily aggregate all the exclude sources (config, user, and  there'll quite likely be module excludes?) when they need to be applied. And there could also be getters for the config/module excludes.

The API you have now seems quite cumbersome, I feel having all the excludes stored in a single set can be limiting, as you can't work with them independently.

m-blaha commented:
Regarding the methods I was following those from `include/libdnf/common/sack/sack.hpp`. Btw is this still a thing? I do not see any subclass of this generic Sack class. May be it's some dead end from previous library designs and could be removed?

I guess modular excludes will be stored separately and will be always generated automatically. I  could be mistaken but I do not see any use case for "user" modular excludes.

I'm not sure about the necessity to separate configuration / "user" excludes - do you have any particular use case in mind?

lukash commented:
> Regarding the methods I was following those from `include/libdnf/common/sack/sack.hpp`. Btw is this still a thing? I do not see any subclass of this generic Sack class. May be it's some dead end from previous library designs and could be removed?

The comps sacks inherit it and also the transaction history sack, maybe there are others. Since it only defins the excludes/includes interface, it doesn't make any sense for the transaction history, I'll quite certainly remove it there. I don't think it's useful for the comps sacks either. I'd still aim for removal unless we want the shared excludes/includes API on some (more than one) classes.

> I guess modular excludes will be stored separately and will be always generated automatically. I could be mistaken but I do not see any use case for "user" modular excludes.

No use and I didn't intend to propose that, it'd be 3 types of excludes "config", "user" and "modular".

> I'm not sure about the necessity to separate configuration / "user" excludes - do you have any particular use case in mind?

We can discuss it tomorrow... My idea is basically to work with the three groups of excludes independently from API user perspective and only merge them for applying to the pool. The reasoning is just that if you store them together, the user can no longer reset one without clearing also the other and having to redo the whole process of adding everything they didn't want to clear (which can arguably be a single call to add config excludes but that's already an ugly nuisance).

m-blaha committed:
libdnf/PackageSack: Split configuration and user excludes

Regular (package) excludes now consist of two sets:

- config_excludes / config_includes - excludes and includes defined in
  main an repository configuration files

- user_excludes / user_includes - excludes and includes created by API
  calls

Those two sets are before applying merged (also with modular excludes)
together.

j-mracek commented:
LGTM! Thank you very much for the contribution.
